ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/InferredTypeHintStage.fs

```diff
@@ -0,0 +1,175 @@
+            if isNull consPat then false else
+
+            // Are we nested inside another ConsPat?
+            let parentTuplePat = TuplePatNavigator.GetByPattern consPat
+            if isNull parentTuplePat then tuplePat.Patterns.IndexOf pattern <> 0 else
+            isNotNull (ConsPatNavigator.GetByPattern1 parentTuplePat)
+
+        else
+
+        //  v-- not evident
+        // [x; y; z]
+        //     ^--^-- evident
+        let listOrListPat = ArrayOrListPatNavigator.GetByPattern pattern
+        if isNotNull listOrListPat then
+            listOrListPat.Patterns.IndexOf pattern <> 0
+
+        else
+
+        false
+
+    override x.VisitNode(node, context) =
+        for child in node.Children() do
+            match child with
+            | :? IFSharpTreeNode as treeNode -> treeNode.Accept(x, context)
+            | _ -> ()
+
+    override x.VisitLocalReferencePat(localRefPat, consumer) =
+        let pat = localRefPat.IgnoreParentParens()
+        if isNotNull (TypedPatNavigator.GetByPattern(pat)) then () else
+
+        let binding = BindingNavigator.GetByHeadPattern(pat)
+        if isNotNull binding && isNotNull binding.ReturnTypeInfo then () else
+
+        if isTypeOfPatternEvident pat then () else
+
+        match box (localRefPat.GetFSharpSymbolUse()) with
+        | null -> ()
+        | symbolUse ->
+
+        let symbolUse = symbolUse :?> FSharpSymbolUse
+        match symbolUse.Symbol with
+        | :? FSharpMemberOrFunctionOrValue as mfv when not (isEvidentFromVariableName mfv.FullType localRefPat.ReferenceName.Identifier.Name) ->
+            let typeNameStr =
+                symbolUse.DisplayContext.WithShortTypeNames(true)
+                |> mfv.FullType.Format
+
+            let range = localRefPat.GetNavigationRange().EndOffsetRange()
+
+            // todo: TypeNameHintHighlighting can be used when RIDER-39605 is resolved
+            consumer.AddHighlighting(TypeHintHighlighting(typeNameStr, range))
+        | _ -> ()
+
+type InferredTypeHintHighlightingProcess
+        (fsFile, settings: IContextBoundSettingsStore, highlightingContext: TypeNameHintHighlightingContext,
+         namingManager: NamingManager, nameParser: NameParser, daemonProcess) =
+    inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
+
+    let namingPolicyProvider = namingManager.Policy.GetPolicyProvider(fsFile.Language, fsFile.GetSourceFile())
+    let hideHintsForEvidentTypes = highlightingContext.HideTypeNameHintsForImplicitlyTypedVariablesWhenTypeIsEvident
+
+    let visitor = LocalReferencePatternVisitor(fsFile, highlightingContext, namingPolicyProvider, nameParser)
+
+    let visitLetBindings (letBindings: ILetBindings) consumer =
+        if not highlightingContext.ShowTypeNameHintsForImplicitlyTypedVariables then () else
+
+        for binding in letBindings.Bindings do
+            if hideHintsForEvidentTypes && isTypeEvident binding.Expression then () else
+            binding.HeadPattern.Accept(visitor, consumer)
```

Comment:
Please check that the pattern is not `null` (which could be true when, for instance, there're errors in code).